async.go

```diff
@@ -160,7 +161,34 @@ func getQueryResultWithRetriesForAsyncMode(
 			logger.WithContext(ctx).Errorf("failed to decode JSON. err: %v", err)
 			return respd, err
 		}
-		if respd.Code != queryInProgressAsyncCode {
+		if respd.Code == sessionExpiredCode {
+			// Update the session token in the header and retry
+			token, _, _ := sr.TokenAccessor.GetTokens()
+			if headers[headerAuthorizationKey] != fmt.Sprintf(headerSnowflakeToken, token) {
+				headers[headerAuthorizationKey] = fmt.Sprintf(headerSnowflakeToken, token)
+				logger.WithContext(ctx).Info("Session token has been updated.")
+				retry++
+				continue
+			}
+
+			// Renew the session token
+			if err = sr.renewExpiredSessionToken(ctx, timeout, token); err != nil {
+				logger.WithContext(ctx).Errorf("failed to renew session token. err: %v", err)
+				return respd, err
+			}
+			retryCountForSessionRenewal++
+
+			// If this is the first response, go back to retry the query
+			// since it failed due to session expiration
+			logger.WithContext(ctx).Infof("retry count for session renewal: %v", retryCountForSessionRenewal)
+			if retryCountForSessionRenewal < 2 {
```

Comment:
Do we need this bumps at all? Maybe it's enough to assume why attempt of retry, so everything below line 178 can be removed? Especially that it is already retried on HTTP retry level. Why do we need this retries in your opinion?

Reply:
Everything below line 178 is used for checking token expiry in normal case where the token expires during `getQueryResultWithRetriesForAsyncMode()`